### ISSUE TITLE
InfluxDB 2.5.1: poor client-side Dashboards performance

### ISSUE THREAD

**Issue by koitsu** [influxdata/ui#6307]
## About the bug

**Steps to reproduce:**
List the minimal actions needed to reproduce the behavior.

1. Visit InfluxDB main UI interface
2. Choose Dashboards
3. Select a dashboard
4. Experience massive client-side slowdown

I did not have these problems under 2.4.0 (that I can recall); in fact, in general 2.x series has been extremely performant (server-side and client-side), so this is the first I've run into performance-related problems.

From a quick diagnostic session in Chrome via Developer Tools, the Dashboard page is issuing hundreds upon hundreds of PATCH requests for no reason I can discern.  Reviewing Performance data shows a huge amount of time in some kind of deeper internal JS-based task, which I suspect is related to the PATCH requests.  From client Console:

```
5 Fetch finished loading: GET "<URL>".
31.989a641181.js:2 [Violation] 'setTimeout' handler took 135ms
15 Fetch finished loading: POST "<URL>".
282 Fetch finished loading: PATCH "<URL>".
31.989a641181.js:2 [Violation] 'click' handler took 669ms
31.989a641181.js:2 [Violation] 'setInterval' handler took 65ms
31.989a641181.js:2 [Violation] 'click' handler took 706ms
31.989a641181.js:2 [Violation] 'setInterval' handler took 73ms
```

Server itself is performing perfectly; CPU load is virtually nothing, memory (RES/RSS and VIRT/VSZ) are reasonable and expected given the size of the dataset.  I can provide `top` output easily, but if you need to know dataset specs, please tell me what I need to provide (`influxd` commands etc.):
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  91518 jdc       20   0 1497228 329056  84360 S   0.7   2.0  19:25.35 influxd
```
My on-disk dataset usage is around 240MBytes.

Important details:
- If I enter a specific board using the cogwheel and choose Configure, the problem DOES NOT happen when editing the board.  In other words: issue only seems to happen when viewing all the dashboards at http://xxx:8086/orgs/xxx/dashboards/xxx?lower=now%28%29+-+7d and not http://xxx:8086/orgs/xxx/dashboards/xxx/cells/xxx/edit
- If after editing a dashboard I go back to the Dashboards page, the problem usually goes away (not always, but usually)
- Amount of lag/stuttering is directly proportional to window period size (i.e. larger window period size = more lag)
- Problem happens in Incognito mode as well

**Expected behavior:**
Performant and not laggy :-) It used to be this way...

**Actual behavior:**
Extreme lag of all sorts.  If you let it run long enough, the browser begins to behave erratically.

**Visual Proof:**
I can try to provide this if necessary; just ask and I can put an .mp4 up somewhere.

## About your environment

**Environment info:**

System info: Linux 5.4.0-131-generic x86_64
InfluxDB version: InfluxDB v2.5.1 (git: 5b6fdbf05d) build_date: 2022-11-02T18:06:28Z
Other relevant environment details: Using official Docker image per DockerHub:
```
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
influxdb     latest    1822a7b80b80   8 days ago    377MB
```
Client: Chrome Version 107.0.5304.107 (Official Build) (64-bit)

InfluxDB server runs on same LAN as client (i.e. no Internet traffic); all connections are wired (i.e. no WiFi).

**Config:**
Stock configuration per `influxd print-config`, with the following changes:
```
no-tasks: true
reporting-disabled: true
storage-tsm-use-madv-willneed: true
```

I can provide any other technical details as asked (I'm a UNIX sysadmin, so server-side stuff tends to be my forte; client-side not so much).

**Performance:**
Performance problems are client-side, not server-side, so `curl` analysis etc. is irrelevant.

**Comment by koitsu** [comment]
I've narrowed down the problem a bit.  Referring to DockerHub image tags and their image IDs:

- `influxdb:latest` image ID `1822a7b80b80` -- slowdown occurs (lots of PATCH requests)
- `influxdb:2.5.0` image ID `973a0d624270` -- slowdown occurs (lots of PATCH requests)
- `influxdb:2.4.0` image ID `6b0ecb57e1f6` -- no problem (no PATCH requests)

TL;DR -- Issue seems to be something introduced between 2.4.0 and 2.5.0.

Given that [the DockerHub tags](https://hub.docker.com/_/influxdb/tags) refer to InfluxDB version numbers and not just the UI part, I am unable to determine which commits in the [influxdb/ui](https://github.com/influxdata/ui) repo might be responsible for the problem.

For testing 2.4.0 vs 2.5.0 vs. 2.5.1, I DID downgrade my DB metadata to 2.4 (via `influxd downgrade 2.4`).  The full process was actually this (where `~/inflxudb/` houses the data currently using 2.5.1):

```
rsync -aH ~/influxdb/ ~/influxdb240/
docker run --rm -v ~/influxdb240/data:/var/lib/influxdb2 -v ~/influxdb240/config:/etc/influxdb2 influxdb:latest influxd downgrade 2.4
vi influxdb240/config/config.yml
#
# Changed http-bind-address to a different port, to not conflict with default of 8086
#
docker pull influxdb:2.4.0
docker run --detach --init --name influxdb240 -v ~/influxdb240/data:/var/lib/influxdb2 -v ~/influxdb240/config:/etc/influxdb2 influxdb:2.4.0
```

I did the same process for 2.5.0 (but without the need for `influxd downgrade 2.4` of course, and used different pathnames to not share backing data storage).

**Comment by koitsu** [comment]
Release notes for InfluxDB 2.5.0: https://github.com/influxdata/influxdb/releases/tag/v2.5.0

Possibly https://github.com/influxdata/influxdb/commit/f36646d is what's responsible?  Sadly I don't have a way of testing 2.5.1 but with the OSS-2022-09-01 UI release (I am not going to build InfluxDB from source :-) ).

**Comment by koitsu** [comment]
Possibly related report, unsure (I'm using Chrome, not Safari): https://github.com/influxdata/influxdb/issues/23870

**Comment by koitsu** [comment]
> @koitsu thanks for the bug ticket. Are the `PATCH` requests happening every time you load a dashboard, or just the first time?

Step-by-step explanation of what happens on 2.5.1, starting with a fresh browser launch (just to ensure a clean slate), and on Chrome 108:

1. Visit InfluxDB URL; for sake of example let's say it's `http://influxdb:8086`
2. Sign in
3. On the left, click Dashboards.  This takes you to `http://influxdb:8086/orgs/XXX/dashboards-list`
4. Click a dashboard; for sake of example let's say it's Test, and that the dashboard has 5 graphs, defaulting to 7 days view.  This takes you to `http://influxdb:8086/orgs/XXX/dashboards/YYY?lower=now%28%29+-+7d`
5. The madness begins here.  Browser begins issuing tons of PATCH requests.  UI is therefore slow as a dog, and can eventually lead to serious problems (as other users describe) depending on how long you let the browser sit here.  Opening up Developer Tools -> Console shows hundreds of PATCH requests.  The `<body>` class is constantly being updated (I can probably track down what exact `<div>` but it's irrelevant and I do not want to deal with front-end if I can avoid it).  CPU load on my system (Intel Core i5-7600K desktop system running Windows 7 SP1 x64) reaches 40% almost instantly.  If I let the browser sit for 10-15 seconds, CPU load hits 82%.  By this point we're at almost 300 PATCH requests.  (If I close the browser or tab at this point, I should note it takes a good 3-4 full seconds before the tab/browser closes, which is quite unusual for any web application.)

Now let's continue:

6. Click the gear icon on a graph and choose Configure.  This takes you to `http://influxdb:8086/orgs/XXX/dashboards/YYY/cells/YYY/edit`
7. The madness stops.  Performance is good; no PATCH requests are happening here
8. Click the Back button in your browser.  You're now back at the URL in Step 4
9. The madness in Step 5 no longer happens; no PATCH requests, performance is good

> Are the `PATCH` requests made to a route that is of the form: `/dashboards/:dashboardId/cells/:cellId/view`?

The PATCH requests seen during Step 5 are for `http://influxdb:8086/api/v2/dashboards/XXX/cells/YYY/view`

> Are the number of `PATCH` requests roughly consistent with the number of line graph cells you have in the dashboard? (Or possibly the total number of line graph series in the dashboard?)

I have absolutely no idea what you mean by this.  What is a "line graph cell"?  I think you're asking me about my data set?  If so, provide me an influx CLI command that will get you the numbers you want and I will issue it.

> If you create a new graph or several new graphs and add them to a dashboard, do you still see the explosion of `PATCH` requests?

I have not tried, as I have no reason to do this.  I feel like this question is more for whoever QAs this stuff, since this should be easily reproducible.

> My hunch is this is related to a feature we call color mapping. The intent of the feature is to provide consistency of line graph series colors across page refreshes. Before we landed the feature, if you had a line graph with two or more series, there was no guarantee that each line's color would remain consistent after a page refresh, which made comprehending the graph an ordeal.

When was this feature landed?  Please provide a version number and/or commit, preferably in the InfluxDB repo itself (and not the UI repo), or maybe both.

> To work around some technical limitations of the graph system in the implementation of color mapping, we have the front end determine the color of each line in a line graph and then update the dashboard cell with a `PATCH` request to save them to disk. This `PATCH` should only happen once - in the ideal case when the graph is created. For graph cells created before this feature existed, these `PATCH` requests should happen the first time the dashboard cell is loaded and not happen again. If the `PATCH`s keep happening, then there's a bug in the system, or this is something else.

If I understand you correctly, then this sounds like a bug in the system.

I think it would be worth you/InfluxDB UI folks reproducing this problem.  What can I provide you to help reproduce the issue?  If you want my data sets and/or dashboard configs, I'll be happy to provide them -- there is nothing sensitive in either of them. What I'll need is the influx CLI commands to get you what you need; I can put up a tarball of the results somewhere online.

**Comment by jeffreyssmith2nd** [comment]
@hoorayimhelping This is due to the graph color mapping. I tracked it down to this [PR](https://github.com/influxdata/ui/pull/5233). We have rolled back to a previous commit in 2.6.0 of OSS to resolve this problem in the short term.

**Comment by koitsu** [comment]
I can confirm that 2.6.0 resolves this problem.  (Specifics: since I use Docker, I'm using `influxdb:latest`, which right now is image ID `df1c226db87e`). Thank you @jeffreyssmith2nd !

**Comment by jeffreyssmith2nd** [comment]
Glad to hear it! I'm reopening as the issue is still present in the UI code, we simply pinned to a commit before the bug was introduced to get things working again.

**Comment by koitsu** [comment]
I have not seen this in quite some time (across several releases/upgrades).  Currently running 2.7.5 (via official Docker image) and things still perform great.  @jeffreyssmith2nd Are you comfortable with this being closed?
